addons/survey/controllers/main.py

```diff
@@ -726,39 +780,30 @@ def _get_user_input_domain(self, survey, line_filter_domain, **post):
+
+        # Create the big fat domain to collect all the user_input_lines matching the filters independently
+        input_line_domain = ['&', ('user_input_id', 'in', user_inputs.ids), ('suggested_answer_id', 'in', suggested_answer_ids)]
+        for domain in filters_domain:
+            input_line_domain = expression.OR([input_line_domain, domain])
+        results = request.env['survey.user_input.line'].search_read(
+            input_line_domain,
+            fields=['id', 'user_input_id', 'question_id', 'answer_type', 'suggested_answer_id', 'value_char_box', 'matrix_row_id']
+        )
+
+        # Then, Keep only the user_input_lines that match all the filters
+        input_line_by_answer_id = defaultdict(list)
+        for result in results:
+            # Filtered Domain
+            if not result['suggested_answer_id']:
+                input_line_by_answer_id[result['question_id'][0]].append(result['user_input_id'][0])
+            # Suggested Answer Id
             else:
-                if row_id and answer_id:
-                    line_filter_domain = expression.AND([
-                        ['&', ('matrix_row_id', '=', row_id), ('suggested_answer_id', '=', answer_id)],
-                        line_filter_domain
-                    ])
-                    answers = request.env['survey.question.answer'].browse([row_id, answer_id])
-                elif answer_id:
-                    line_choices.append(answer_id)
-                    answers = request.env['survey.question.answer'].browse([answer_id])
-                if answer_id:
-                    question_id = answers[0].matrix_question_id or answers[0].question_id
-                    search_filters.append({
-                        'row_id': row_id,
-                        'answer_id': answer_id,
-                        'question': question_id.title,
-                        'answers': '%s%s' % (answers[0].value, ': %s' % answers[1].value if len(answers) > 1 else '')
-                    })
-        if line_choices:
-            line_filter_domain = expression.AND([[('suggested_answer_id', 'in', line_choices)], line_filter_domain])
+                input_line_by_answer_id[result['suggested_answer_id'][0]].append(result['user_input_id'][0])
+        filtered_user_input_ids = set.intersection(*map(set, input_line_by_answer_id.values())) if input_line_by_answer_id else input_line_by_answer_id
 
-        user_input_domain = self._get_user_input_domain(survey, line_filter_domain, **post)
-        user_input_lines = request.env['survey.user_input'].sudo().search(user_input_domain).mapped('user_input_line_ids')
+        # Finally, get the user input line records matching all the filters, or every user input line records if no filters
+        user_input_line_records = user_inputs.filtered_domain([('id', 'in', list(filtered_user_input_ids))]).mapped('user_input_line_ids') \
+            if filtered_user_input_ids else user_inputs.mapped('user_input_line_ids')
```

Comment:
if you have all the user input ids, you can directly browse them instead of applying a domain.
```
if filtered_user_input_ids:
    user_inputs = self.env['survey.user_input'].browse(filtered_user_input_ids)
return user_inputs.user_input_line_ids, filters_data
```
but now that I think of it, we can use the  search read to retrieve all the user_input _lines linked to "common user_inputs". So we won't need to read the recordset but the dict instead.
`input_line_ids = ... #get it from results`
then finally `return request.env['survey.user_input.line'].browse(input_line_ids), filters_data`


EDIT: All in all after all my comments... this would become something like:
```
if input_line_by_answer_id:
    filtered_user_input_ids = set.intersection(*map(set, input_line_by_answer_id.values()))
    user_input_line_ids = [result['id'] for result in results if result['user_input_id'] in filtered_user_input_ids]
    user_input_lines = self.env['survey.user_input.line'].browse(user_input_line_ids)
else:
    user_input_lines = user_inputs.user_input_line_ids
return user_input_lines, filters_data
```
Didn't test it but that's the idea.

Reply:
It was a good idea however the `result['id']` value is only the `user_input_line_id` of the filter instead of all the `user_input_line_ids` of the `user_input` matching the filter. This is because the `results` is a dict containing only the information about the `user_input_lines` matching the filters **independently**.

To demonstrate, if we filter on 'Brussels', its `user_input_line_id` will be for example `189`. So when you're returning the `user_input_lines` you will return only the  `user_input_line` record of the `Brussels` answer and all the other questions will just show a `No answered yet` helper. 

Also I just realized that the `input_line_by_answer_id` isn't really a list of `user_input_line_id` but a list of `user_input_id` as I "append" `result['user_input_id'][0]`.


So we know that `user_inputs` is all the user inputs of the survey, we also have a list of `user_input_ids` matching the filters, previously called  `input_line_by_answer_id` that I'll rename as `user_input_by_answer_id`. So why don't we just get the matching records from the `user_inputs` ?

Here is my code:

```python
        user_inputs = request.env['survey.user_input'].sudo().search(user_input_domain)

        # Create the domain to collect all the user_input_lines matching the filters independently
        input_line_domain = ['&', ('user_input_id', 'in', user_inputs.ids), ('suggested_answer_id', 'in', suggested_answer_ids)]
        for domain in filters_domain:
            input_line_domain = expression.OR([input_line_domain, domain])
        results = request.env['survey.user_input.line'].search_read(
            input_line_domain,
            fields=['id', 'user_input_id', 'question_id', 'answer_type', 'suggested_answer_id', 'value_char_box', 'matrix_row_id']
        )
        # Then, keep only the user input ids that match all the filters
        user_input_by_answer_id = defaultdict(list)
        for result in results:
            # Filtered Domain
            if not result['suggested_answer_id']:
                user_input_by_answer_id[result['question_id'][0]].append(result['user_input_id'][0])
            # Suggested Answer Id
            else:
                user_input_by_answer_id[result['suggested_answer_id'][0]].append(result['user_input_id'][0])

        # Finally, get all the user input lines of the matching user input records
        if user_input_by_answer_id:
            filtered_user_input_ids = set.intersection(*map(set, user_input_by_answer_id.values()))
            user_inputs = user_inputs.browse(filtered_user_input_ids)
        user_input_lines = user_inputs.user_input_line_ids

        return user_input_lines, filters_data
```

I didn't push yet (waiting for your opinion) but feel free to tell me if it's necessary to test it :)